src/libraries/System.Private.CoreLib/src/System/Guid.cs

```diff
@@ -1324,133 +1312,193 @@ private bool TryFormatCore<TChar>(Span<TChar> destination, out int charsWritten,
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)] // only used from two callers
-        internal unsafe bool TryFormatCore<TChar>(Span<TChar> destination, out int charsWritten, int flags) where TChar : unmanaged, IUtfChar<TChar>
+        internal bool TryFormatCore<TChar>(Span<TChar> destination, out int charsWritten, int flags)
+            where TChar : unmanaged, IUtfChar<TChar>
         {
+            Span<TChar> dest = destination;
+
             // The low byte of flags contains the required length.
-            if ((byte)flags > destination.Length)
+            if ((byte)flags > dest.Length)
             {
                 charsWritten = 0;
                 return false;
             }
 
-            charsWritten = (byte)flags;
+            int length = (byte)flags;
+            charsWritten = length;
+
             flags >>= 8;
+            if ((byte)flags != 0)
+            {
+                dest[0] = TChar.CastFrom((byte)flags);
+                dest[length - 1] = TChar.CastFrom((byte)(flags >> 8));
+                dest = dest.Slice(1);
+            }
 
-            fixed (TChar* guidChars = &MemoryMarshal.GetReference(destination))
+            if (Ssse3.IsSupported || AdvSimd.Arm64.IsSupported)
+            {
+                FormatCore_Vector(dest, (flags >> 8) < 0 /* dash */);
+            }
+            else
             {
-                TChar* p = guidChars;
+                FormatCore_Scalar(dest, (flags >> 8) < 0 /* dash */);
+            }
+            return true;
+        }
 
-                // The low byte of flags now contains the opening brace char (if any)
-                if ((byte)flags != 0)
-                {
-                    *p++ = TChar.CastFrom((byte)flags);
-                }
-                flags >>= 8;
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        [CompExactlyDependsOn(typeof(Ssse3))]
+        [CompExactlyDependsOn(typeof(AdvSimd.Arm64))]
+        internal void FormatCore_Vector<TChar>(Span<TChar> destination, bool hasDashes) where TChar : unmanaged, IUtfChar<TChar>
+        {
+            Span<TChar> dest = destination;
 
-                if ((Ssse3.IsSupported || AdvSimd.Arm64.IsSupported) && BitConverter.IsLittleEndian)
-                {
-                    // Vectorized implementation for D, N, P and B formats:
-                    // [{|(]dddddddd[-]dddd[-]dddd[-]dddd[-]dddddddddddd[}|)]
-                    (Vector128<byte> vecX, Vector128<byte> vecY, Vector128<byte> vecZ) = FormatGuidVector128Utf8(this, flags < 0 /* dash */);
+            // Vectorized implementation for D, N, P and B formats:
+            // [{|(]dddddddd[-]dddd[-]dddd[-]dddd[-]dddddddddddd[}|)]
+            (Vector128<byte> vecX, Vector128<byte> vecY, Vector128<byte> vecZ) = FormatGuidVector128Utf8(this, hasDashes);
```

Comment:
It's probably possible to make this fully xplat now and then we could remove the CompExactlyDependsOn